File: particles/Tutorial/Javascript/Demo/source/TTTGame.js
/**
 * @license
 * Copyright 2019 Google LLC.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

'use strict';

/* global defineParticle */
defineParticle(({SimpleParticle, html}) => {

  const winningSequences = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6]
  ];

  const template = html`
It is your turn <span>{{playerDetails}}</span>.
<div slotid="gameSlot"></div>
<div hidden="{{hideCongrats}}">Congratulations <span>{{winnerName}}</span>!</div>`;

  return class extends SimpleParticle {

    get template() {
      return template;
    }

    shouldRender({gameState, playerOne, playerTwo}) {
      return gameState && playerOne.id === 0 && playerTwo.id === 1;
    }

    update({gameState, playerOne, playerTwo, playerOneMove, playerTwoMove, events}) {

      if (playerOne.id != 0) {
        this.set('playerOne', {name: playerOne.name, avatar: playerOne.avatar, id: 0});
      }

      if (!playerTwo.id) {
        this.set('playerTwo', {name: playerTwo.name, avatar: playerTwo.avatar, id: 1});
      }

      if (!gameState) {
        this.newGame();
      }

      if (gameState && gameState.gameOver && events) {
        const reset = events.find(e => e.type === 'reset');
        if (reset) {
          this.newGame();
        }
      }

      if (gameState && !gameState.gameOver) {
        const mv = [playerOneMove.move, playerTwoMove.move][gameState.currentPlayer];
        const avatar = [playerOne.avatar, playerTwo.avatar][gameState.currentPlayer];
        // Create a local mutable copy of gameState to manipulate.
        let gs = {...gameState};
        gs.board = JSON.parse(gs.board);
        if (this.isMoveValid(gs, mv)) {
          gs = this.applyMove(gs, mv, avatar);
          gs = this.applyGameStatus(gs, playerOne, playerTwo);
          gs.board = JSON.stringify(gs.board);
          this.set('gameState', gs);
        }
      }
    }

    render({gameState, playerOne, playerTwo}) {
      return {
        hideCongrats: !gameState.gameOver,
        playerDetails: this.getPlayerDetails(gameState.currentPlayer, playerOne, playerTwo),
        winnerName: gameState.winnerId !== null ? this.getPlayerDetails(gameState.winnerId, playerOne, playerTwo) : `it's a tie`
      };
    }

    newGame() {
      this.set('playerOneMove', {});
      this.set('playerTwoMove', {});
      this.clear('events');
      this.set('gameState', {
        board: JSON.stringify(['', '', '', '', '', '', '', '', '']),
        gameOver: false,
        winnerId: null,
        currentPlayer: Math.floor(Math.random() * 2)
      });
    }

    isMoveValid(gs, mv) {
      return mv > -1 && mv < 10 && gs.board[mv] === '';
    }

    applyMove(gs, mv, avatar) {
      gs.board[mv] = avatar;
      gs.currentPlayer = (gs.currentPlayer + 1) % 2;
      return gs;
    }

    applyGameStatus(gs, p1, p2) {
      // Check if the game is tied
      gs.gameOver = gs.board.every(cell => cell !== '');

      // Check if the game has been won
      for (const ws of winningSequences) {
        if (gs.board[ws[0]] !== '' && gs.board[ws[0]] === gs.board[ws[1]] && gs.board[ws[1]] === gs.board[ws[2]]) {
          gs.gameOver = true;
          gs.winnerId = gs.board[ws[0]] === p1.avatar ? p1.id : p2.id;
          break;
        }
      }
      return gs;
    }

    getPlayerDetails(playerId, p1, p2) {
      const name = [p1.name, p2.name][playerId];
      const avatar = [p1.avatar, p2.avatar][playerId];
      return `${name} playing as ${avatar}`;
    }
  };
});

File: particles/Tutorial/Kotlin/Demo/src/TTTGame.kt
/*
 * Copyright 2019 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.tutorials.tictactoe

import arcs.sdk.Handle

class TTTGame : AbstractTTTGame() {
    private val defaultGame = TTTGame_GameState(
        board = ",,,,,,,,",
        currentPlayer = (0..1).random().toDouble(),
        gameOver = false,
        winnerAvatar = ""
    )

    private val winningSequences = arrayOf(
        arrayOf(0, 1, 2),
        arrayOf(3, 4, 5),
        arrayOf(6, 7, 8),
        arrayOf(0, 3, 6),
        arrayOf(1, 4, 7),
        arrayOf(2, 5, 8),
        arrayOf(0, 4, 8),
        arrayOf(2, 4, 6)
    )

    init {

        handles.playerOneMove.onUpdate { move ->
            applyMove(
              move ?: TTTGame_PlayerOneMove(),
              handles.playerOne.fetch() ?: TTTGame_PlayerOne()
            )
        }

        handles.playerTwoMove.onUpdate { move ->
            applyMove(
              move ?: TTTGame_PlayerTwoMove(),
              handles.playerTwo.fetch() ?: TTTGame_PlayerTwo()
            )
        }

        handles.events.onUpdate {
            if (hasReset()) {
                with(handles) {
                    gameState.set(defaultGame)
                    playerOneMove.set(handles.playerOneMove.fetch()!!.copy(move = -1.0))
                    playerTwoMove.set(handles.playerTwoMove.fetch()!!.copy(move = -1.0))
                    events.clear()
                }
                renderOutput()
            }
        }
    }

    private fun applyMove(move: TTTGame_PlayerOneMove, player: TTTGame_PlayerOne) {
        val mv = move.move.toInt()
        val gs = handles.gameState.fetch() ?: TTTGame_GameState()

        // Check if move is valid
        val boardList = gs.board.split(",").toMutableList()
        if (!mv.isValidMove(boardList) || gs.currentPlayer != player.id || gs.gameOver) return

        // Apply the move
        boardList[mv] = player.avatar

        var gameOver = !boardList.contains("")
        var avatar = ""

        // Check if the game is over
        winningSequences.forEach { sequence ->
            if (boardList[sequence[0]] != "" &&
              boardList[sequence[0]] == boardList[sequence[1]] &&
              boardList[sequence[0]] == boardList[sequence[2]]) {
                gameOver = true
                avatar = player.avatar
            }
        }

        handles.gameState.set(gs.copy(
          board = boardList.joinToString(","),
          currentPlayer = (gs.currentPlayer + 1) % 2,
          gameOver = gameOver,
          winnerAvatar = avatar
        ))
        renderOutput()
    }

    override fun onHandleSync(handle: Handle, allSynced: Boolean) {
        if (handles.gameState.fetch()?.board == null) {
            handles.gameState.set(defaultGame)
        }
        if (handle.name == "playerOne" && handles.playerOne.fetch()?.id != 0.0) {
            val p1 = handles.playerOne.fetch() ?: TTTGame_PlayerOne()
            handles.playerOne.set(p1.copy(id = 0.0))
        }
        if (handle.name == "playerTwo" && handles.playerTwo.fetch()?.id != 1.0) {
            val p2 = handles.playerTwo.fetch() ?: TTTGame_PlayerTwo()
            handles.playerTwo.set(p2.copy(id = 1.0))
        }
    }

    override fun populateModel(slotName: String, model: Map<String, Any>): Map<String, Any> {
        val gs = handles.gameState.fetch() ?: TTTGame_GameState()
        val p1 = handles.playerOne.fetch() ?: TTTGame_PlayerOne()
        val p2 = handles.playerTwo.fetch() ?: TTTGame_PlayerTwo()

        val cpName = if (gs.currentPlayer == p1.id) p1.name else p2.name
        val cpAvatar = if (gs.currentPlayer == p1.id) p1.avatar else p2.avatar

        val congratsMessage = gs.winnerAvatar.let {
            if (it == p1.avatar) "Congratulations ${p1.name}, you won!"
            else if (it == p2.avatar) "Congratulations ${p2.name}, you won!"
            else "It's a tie!"
        }

        return mapOf(
            "message" to congratsMessage,
            "hideCongrats" to !gs.gameOver,
            "playerDetails" to "$cpName playing as $cpAvatar"
        )
    }

    override fun getTemplate(slotName: String): String = """
        It is your turn <span>{{playerDetails}}</span>.
        <div slotid="boardSlot"></div>
        <div hidden="{{hideCongrats}}"><span>{{message}}</span></div>
        """.trimIndent()

    private fun hasReset() = handles.events.fetchAll().any { it.type == "reset" }

    private fun Int.isValidMove(boardList: List<String>) = this in 0..9 && boardList[this] == ""
}
